File: du-backend/src/main/java/com/dwin/du/api/controller/LoanController.java
package com.dwin.du.api.controller;
import com.dwin.du.api.request.AddEditLoanRequest;
import com.dwin.du.api.service.LoanService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
@RequestMapping("/rm/loan")
@Tag(name = "Loan Management", description = "APIs for managing loans")
public class LoanController {

    private final LoanService loanService;

    @PostMapping("/add")
    @Operation(summary = "Add a new loan", description = "Adds a new loan to the system.")
    public ResponseEntity<?> addLoan(@RequestBody AddEditLoanRequest request){
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return loanService.addLoan(username, request);
    }

    @PutMapping("/{loanId}/edit")
    @Operation(summary = "Edit a loan", description = "Edits the details of an existing loan by its ID.")
    public ResponseEntity<?> editLoan(@PathVariable int loanId, @RequestBody AddEditLoanRequest request) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return loanService.editLoan(username, loanId, request);
    }

    @DeleteMapping("/{loanId}/remove")
    @Operation(summary = "Remove a loan", description = "Removes a loan from the system by its ID.")
    public ResponseEntity<?> removeLoan(@PathVariable int loanId){
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return loanService.removeLoan(username, loanId);
    }

    @PutMapping("/{loanId}/set-settled")
    @Operation(summary = "Set loan as settled", description = "Marks a loan as settled by its ID.")
    public ResponseEntity<?> setIsSettled(@PathVariable int loanId, @RequestParam boolean settled) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return loanService.setIsSettled(username, loanId, settled);
    }

    @PutMapping("/{loanId}/set-lent")
    @Operation(summary = "Set loan as lent", description = "Marks a loan as lent by its ID.")
    public ResponseEntity<?> setIsLent(@PathVariable int loanId, @RequestParam boolean lent) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return loanService.setIsLent(username, loanId, lent);
    }

    @GetMapping("/{loanId}")
    @Operation(summary = "Retrieve a loan", description = "Retrieves the details of a loan by its ID.")
    public ResponseEntity<?> getLoan(@PathVariable int loanId) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return loanService.getLoan(username, loanId);
    }

    @GetMapping("/person/{personId}")
    @Operation(summary = "Retrieve a loan by person", description = "Retrieves the details of a loan associated with a specific person by their ID.")
    public ResponseEntity<?> getLoanPerson(@PathVariable int personId) {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return loanService.getLoanPerson(username, personId);
    }

    @GetMapping("")
    @Operation(summary = "Retrieve all loans", description = "Retrieves all loans associated with the current user.")
    public ResponseEntity<?> getLoans(){
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = authentication.getName();
        return loanService.getLoans(username);
    }
}

File: divvyup/DivvyUp.Web/Controllers/LoanController.cs
using DivvyUp_Shared.AppConstants;
using DivvyUp_Shared.Dtos.Request;
using DivvyUp_Shared.Interfaces;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Swashbuckle.AspNetCore.Annotations;

namespace DivvyUp.Web.Controllers
{
    [Authorize]
    [ApiController]
    [SwaggerTag("Loan Management")]
    [Route(ApiRoute.LOAN_ROUTES.LOAN_ROUTE)]
    public class LoanController : ControllerBase
    {
        private readonly ILoanService _loanService;

        public LoanController(ILoanService loanService)
        {
            _loanService = loanService;
        }

        [HttpPost(ApiRoute.LOAN_ROUTES.ADD)]
        [SwaggerOperation(Summary = "Add a new loan", Description = "Adds a new loan to the system.")]
        public async Task<IActionResult> Add([FromBody] AddEditLoanDto request)
        {
           await _loanService.Add(request);
           return Ok();
        }

        [HttpPut(ApiRoute.LOAN_ROUTES.EDIT)]
        [SwaggerOperation(Summary = "Edit a loan", Description = "Edits the details of an existing loan by its ID.")]
        public async Task<IActionResult> Edit([FromBody] AddEditLoanDto request, [FromRoute] int loanId)
        {
            await _loanService.Edit(request, loanId);
            return Ok();
        }

        [HttpDelete(ApiRoute.LOAN_ROUTES.REMOVE)]
        [SwaggerOperation(Summary = "Remove a loan", Description = "Removes a loan from the system by its ID.")]
        public async Task<IActionResult> Remove([FromRoute] int loanId)
        {
            await _loanService.Remove(loanId);
            return Ok();
        }

        [HttpPut(ApiRoute.LOAN_ROUTES.SET_PERSON)]
        [SwaggerOperation(Summary = "Set person in loan", Description = "Changes the person in an existing loan.")]
        public async Task<IActionResult> SetPerson([FromRoute] int loanId, [FromRoute] int personId)
        {
            await _loanService.SetPerson(loanId, personId);
            return Ok();
        }

        [HttpPut(ApiRoute.LOAN_ROUTES.SET_SETTLED)]
        [SwaggerOperation(Summary = "Set loan as settled", Description = "Marks a loan as settled by its ID.")]
        public async Task<IActionResult> SetSettled([FromRoute] int loanId, [FromRoute] bool settled)
        {
            await _loanService.SetSettled(loanId, settled);
            return Ok();
        }

        [HttpPut(ApiRoute.LOAN_ROUTES.SET_LENT)]
        [SwaggerOperation(Summary = "Set loan as lent", Description = "Marks a loan as lent by its ID.")]
        public async Task<IActionResult> SetLent([FromRoute] int loanId, [FromRoute] bool lent)
        {
            await _loanService.SetLent(loanId, lent);
            return Ok();
        }

        [HttpGet(ApiRoute.LOAN_ROUTES.LOAN)]
        [SwaggerOperation(Summary = "Retrieve a loan", Description = "Retrieves the details of a loan by its ID.")]
        public async Task<IActionResult> GetLoan([FromRoute] int loanId)
        {
            var loan = await _loanService.GetLoan(loanId);
            return Ok(loan);
        }

        [HttpGet(ApiRoute.LOAN_ROUTES.LOANS)]
        [SwaggerOperation(Summary = "Retrieve all loans", Description = "Retrieves all loans associated with the current user.")]
        public async Task<IActionResult> GetLoans()
        {
            var loans = await _loanService.GetLoans();
            return Ok(loans);
        }

        [HttpGet(ApiRoute.LOAN_ROUTES.LOANS_PERSON)]
        [SwaggerOperation(Summary = "Retrieve a loan by person", Description = "Retrieves the details of a loan associated with a specific person by their ID.")]
        public async Task<IActionResult> GetPersonLoans([FromRoute] int personId)
        {
            var loans = await _loanService.GetPersonLoans(personId);
            return Ok(loans);
        }

        [HttpGet(ApiRoute.LOAN_ROUTES.LOANS_DATA_RANGE)]
        [SwaggerOperation(Summary = "Retrieve all loans in date range", Description = "Retrieves all loans in date range associated with the current user.")]
        public async Task<IActionResult> GetLoansByDataRange([FromQuery] DateOnly from, [FromQuery] DateOnly to)
        {
            var loans = await _loanService.GetLoansByDataRange(from, to);
            return Ok(loans);
        }
    }
}
